compiler/resolution/preFold.cpp

```diff
@@ -390,59 +395,55 @@ static void setRecordDefaultValueFlags(AggregateType* at) {
       ts->addFlag(FLAG_TYPE_NO_DEFAULT_VALUE);
     } else if (isNilableClassType(at)) {
       ts->addFlag(FLAG_TYPE_DEFAULT_VALUE);
-
     } else if (Type* eltType = arrayElementType(at)) {
-      if (AggregateType* eltTypeAt = toAggregateType(eltType)) {
-        setRecordAssignableFlags(eltTypeAt);
-
-        if (eltType->symbol->hasFlag(FLAG_TYPE_DEFAULT_VALUE))
-          at->symbol->addFlag(FLAG_TYPE_DEFAULT_VALUE);
-        else if (eltType->symbol->hasFlag(FLAG_TYPE_NO_DEFAULT_VALUE))
-          at->symbol->addFlag(FLAG_TYPE_NO_DEFAULT_VALUE);
+      if (isDefaultInitializable(eltType)) {
+        ts->addFlag(FLAG_TYPE_DEFAULT_VALUE);
       } else {
-         at->symbol->addFlag(FLAG_TYPE_DEFAULT_VALUE);
+        ts->addFlag(FLAG_TYPE_NO_DEFAULT_VALUE);
       }
 
-
     } else if (at->symbol->hasFlag(FLAG_EXTERN)) {
       // Currently extern records aren't initialized at all by default.
       // But it's not necessarily reasonable to expect them to have
       // initializers. See issue #7992.
       ts->addFlag(FLAG_TYPE_DEFAULT_VALUE);
     } else {
+
+      if (at->hasUserDefinedInit == false ||
```

Comment:
Does this code make the code I added in #15747 redundant? It looks like it does.

In that case, why don't I just wait for you to merge this effort and then use your changes?

Reply:
I started with your changes and then made updates.